src/ast/irbuilderbpf.cpp

```diff
@@ -362,15 +362,30 @@ Value *IRBuilderBPF::CreateUSDTReadArgument(Value *ctx, struct bcc_usdt_argument
       abort();
     }
 
+    // bpftrace's args are internally represented as 64 bit integers. However,
+    // the underlying argument (of the target program) may be less than 64
+    // bits. So we must be careful to zero out unused bits.
     Value* reg = CreateGEP(ctx, getInt64(offset * sizeof(uintptr_t)), "load_register");
     AllocaInst *dst = CreateAllocaBPF(builtin.type, builtin.ident);
-    CreateProbeRead(dst, builtin.type.size, reg);
-    result = CreateLoad(dst);
     if (argument->valid & BCC_USDT_ARGUMENT_DEREF_OFFSET) {
+      // If we are dereferencing a pointer for the offset, we want to
+      // read native pointer size (assumed to be 64-bit here)
+      assert(builtin.type.size == 8);
+      CreateProbeRead(dst, builtin.type.size, reg);
+      result = CreateLoad(dst);
```

Comment:
I think it is possible to load reg directly because it's in context memory.

```suggestion
      result = CreateLoad(getInt64Ty(), reg)
```